src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/brgemm_executor.hpp

```diff
@@ -0,0 +1,112 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+#pragma once
+
+#include <cpu/x64/amx_tile_configure.hpp>
+#include <cpu/x64/brgemm/brgemm.hpp>
+#include <cpu/x64/matmul/brgemm_matmul_copy_utils.hpp>
+#include <cpu/x64/matmul/brgemm_matmul_utils.hpp>
+#include <cstddef>
+#include <openvino/core/type/element_type.hpp>
+
+namespace ov {
+namespace intel_cpu {
+
+class brgemmExecutor {
+public:
+    // Construct brgemm kernel for matmul (M, K) * (K, N)/(N, K)^T
+    // BF16 * BF16 -> FP32
+    // lda is the first dimension for A matrix
+    // ldb is the first dimension for B matrix
+    // ldc is the first dimension for C matrix
+    // b_transpose indicates wheter B matrix is transposed.
+    brgemmExecutor(size_t M,
+                   size_t N,
+                   size_t K,
+                   size_t lda,
+                   size_t ldb,
+                   size_t ldc,
+                   bool b_transposed = false);
+    // execute all M
+    void executeGemm(void* a, void* b, void* c, void* wsp, void* scratch_a, void* scratch_b);
+    // execute m_blk
+    void executeGemm(size_t m_blk, void* a, void* b, void* c, void* wsp, void* scratch_a, void* scratch_b);
+
+    void copy_buffer_a();
+    void copy_buffer_b(void* b, void* scratch_b);
+    // bytes needed to place scratch buffer a
+    const size_t get_scratch_a_size() const;
+    // bytes needed to place scratch buffer b
+    const size_t get_scratch_b_size() const;
+    const size_t get_mblk_size() const {
+        return matmulOptimalM;
+    }
+    const size_t get_k_blk() const {
+        return K_blk;
+    }
+    const size_t get_wsp_size() const {
+        return 4 * 1024;
+    }
+
+private:
+    size_t M = 0, M_blk = 0, M_tail = 0;
+    size_t K = 0, K_blk = 0, K_tail = 0, N = 0, N_tail = 0;
+    size_t lda = 0, ldb = 0, ldc = 0;
+    bool b_transposed = false;
+    size_t brgVnniFactor = 0;
+    size_t packedBSize = 0;
+    size_t packedASize = 0;
+    static constexpr size_t N_blk = 32;
+    static constexpr size_t MHA_BRGEMM_KERNELS_NUM = 8;
+    static constexpr size_t matmulOptimalM = 32;
+    struct brgemmCtx {
+        size_t M = 0, N = 0, K = 0, LDA = 0, LDB = 0, LDC = 0;
+        dnnl_data_type_t dt_in0 = dnnl_data_type_undef;
+        dnnl_data_type_t dt_in1 = dnnl_data_type_undef;
+        char palette[64];
+        bool is_with_amx = false;
+        bool is_with_comp = false;
+        bool transpose_a = false;
+        bool transpose_b = false;
+        float beta = 0.0f;
+    };
+    brgemmCtx brgCtxs[MHA_BRGEMM_KERNELS_NUM];
+    std::unique_ptr<dnnl::impl::cpu::x64::brgemm_kernel_t> brgKernels[MHA_BRGEMM_KERNELS_NUM];
```

Comment:
Do we need a destructor to do something like brgemm_kernel_destroy and amx_tile_release?

Reply:
- `brgemm_kernel_destroy` simply deletes the poniter, I search the code of oneDNN no kernel calls it. The memory of brgemm jit kernel is owned by the Brgemm_Executor, after Brgemm_Executor is deleted, that memory is also deleted.

- According to `Optimization Reference Manual`, the `amx_tile_release` is meant to called inside kernel . @dmitry-gorokhov Do you have suggestions?